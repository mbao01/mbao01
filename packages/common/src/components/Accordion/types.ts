import * as AccordionPrimitive from "@radix-ui/react-accordion";

export type AccordionProps = React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Root>;

export type AccordionContentProps = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Content
>;

export type AccordionItemProps = React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Item>;

export type AccordionTriggerProps = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Trigger
>;
