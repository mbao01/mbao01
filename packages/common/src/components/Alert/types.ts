import { type VariantProps } from "../../libs";
import { getAlertClasses } from "./constants";

export type AlertProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getAlertClasses>;

export type AlertTitleProps = React.HTMLAttributes<HTMLHeadingElement>;

export type AlertDescriptionProps = React.HTMLAttributes<HTMLDivElement>;
