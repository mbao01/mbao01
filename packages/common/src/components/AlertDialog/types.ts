import * as AlertDialogPrimitive from "@radix-ui/react-alert-dialog";
import { type VariantProps } from "../../libs";
import { getButtonClasses } from "../Button/constants";
import {
  getDialogContentClasses,
  getDialogDescriptionClasses,
  getDialogFooterClasses,
  getDialogHeaderClasses,
  getDialogOverlayClasses,
  getDialogTitleClasses,
} from "../Dialog/constants";

export type AlertDialogProps = React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Root>;

export type AlertDialogOverlayProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Overlay
> &
  VariantProps<typeof getDialogOverlayClasses>;

export type AlertDialogContentProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Content
> &
  VariantProps<typeof getDialogContentClasses>;

export type AlertDialogHeaderProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogHeaderClasses>;

export type AlertDialogFooterProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogFooterClasses>;

export type AlertDialogTitleProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Title
> &
  VariantProps<typeof getDialogTitleClasses>;

export type AlertDialogDescriptionProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Description
> &
  VariantProps<typeof getDialogDescriptionClasses>;

export type AlertDialogActionProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Cancel
> &
  VariantProps<typeof getButtonClasses>;

export type AlertDialogCancelProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Cancel
> &
  VariantProps<typeof getButtonClasses>;
