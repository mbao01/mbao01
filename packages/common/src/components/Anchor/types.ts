import { VariantProps } from "../../libs";
import { getAnchorClasses } from "./constant";

export type AnchorProps = React.AnchorHTMLAttributes<HTMLAnchorElement> & {
  isExternal?: boolean;
} & VariantProps<typeof getAnchorClasses>;
