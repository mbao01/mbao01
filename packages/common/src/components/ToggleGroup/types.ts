import * as ToggleGroupPrimitive from "@radix-ui/react-toggle-group";
import { BaseToggleProps } from "../Toggle/types";

export type ToggleGroupProps = React.ComponentPropsWithoutRef<
  typeof ToggleGroupPrimitive.Root
> &
  Pick<BaseToggleProps, "variant">;

export type ToggleGroupItemProps = React.ComponentPropsWithoutRef<
  typeof ToggleGroupPrimitive.Item
> &
  BaseToggleProps;

export type ToggleGroupContextType = Pick<BaseToggleProps, "variant">;
