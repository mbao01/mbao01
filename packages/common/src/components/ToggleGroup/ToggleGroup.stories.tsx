import type { Meta, StoryContext, StoryFn, StoryObj } from "@storybook/react";
import { BoldIcon, ItalicIcon, UnderlineIcon } from "lucide-react";
import { ToggleGroup } from "./ToggleGroup";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const withToggleGroup = (_: StoryFn, context: StoryContext<any>) => {
  const { type, variant, defaultValue } = context.args;

  return (
    <ToggleGroup type={type} variant={variant} defaultValue={defaultValue}>
      <ToggleGroup.Item value="bold" aria-label="Toggle bold">
        <BoldIcon className="h-4 w-4" />
      </ToggleGroup.Item>
      <ToggleGroup.Item value="italic" aria-label="Toggle italic">
        <ItalicIcon className="h-4 w-4" />
      </ToggleGroup.Item>
      <ToggleGroup.Item value="strikethrough" aria-label="Toggle strikethrough">
        <UnderlineIcon className="h-4 w-4" />
      </ToggleGroup.Item>
    </ToggleGroup>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: ToggleGroup,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withToggleGroup],
} satisfies Meta<typeof ToggleGroup>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    variant: "default",
    type: "multiple",
  },
};

export const PrimaryToggle: Story = {
  args: {
    variant: "primary",
    type: "multiple",
  },
};

export const SingleToggle: Story = {
  args: {
    type: "single",
    variant: "accent",
  },
};

export const SinglePreselectedToggle: Story = {
  args: {
    type: "single",
    defaultValue: "italic",
  },
};

export const MultiplePreselectedToggle: Story = {
  args: {
    type: "multiple",
    variant: "warning",
    defaultValue: ["italic", "bold"],
  },
};
