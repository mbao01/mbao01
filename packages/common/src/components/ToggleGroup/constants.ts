import { cva } from "../../libs";

export const getToggleGroupClasses = cva("flex items-center justify-center gap-1");
