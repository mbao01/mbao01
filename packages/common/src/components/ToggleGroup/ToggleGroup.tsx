"use client";

import { createContext, forwardRef, useContext } from "react";
import * as ToggleGroupPrimitive from "@radix-ui/react-toggle-group";
import type {
  ToggleGroupContextType,
  ToggleGroupItemProps,
  ToggleGroupProps,
} from "./types";
import { cn } from "../../utilities";
import { getToggleClasses } from "../Toggle/constants";
import { getToggleGroupClasses } from "./constants";
import { getButtonClasses } from "../Button/constants";

const ToggleGroupContext = createContext<ToggleGroupContextType>({
  variant: "default",
});

const ToggleGroup = ({
  className,
  variant,
  children,
  ...props
}: ToggleGroupProps) => (
  <ToggleGroupPrimitive.Root
    className={cn(getToggleGroupClasses(), className)}
    {...props}
  >
    <ToggleGroupContext.Provider value={{ variant }}>
      {children}
    </ToggleGroupContext.Provider>
  </ToggleGroupPrimitive.Root>
);

const ToggleGroupItem = forwardRef<
  React.ElementRef<typeof ToggleGroupPrimitive.Item>,
  ToggleGroupItemProps
>(({ className, children, variant, size, wide, outline, ...props }, ref) => {
  const context = useContext(ToggleGroupContext);

  return (
    <ToggleGroupPrimitive.Item
      ref={ref}
      className={cn(
        getButtonClasses({ size, wide, outline }),
        getToggleClasses({
          variant: context.variant ?? variant,
        }),
        className
      )}
      {...props}
    >
      {children}
    </ToggleGroupPrimitive.Item>
  );
});

ToggleGroupItem.displayName = ToggleGroupPrimitive.Item.displayName;

ToggleGroup.Item = ToggleGroupItem;

export { ToggleGroup };
