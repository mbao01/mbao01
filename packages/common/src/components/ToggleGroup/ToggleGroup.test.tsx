import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { type ToggleGroupProps } from "./types";
import { ToggleGroup } from "./ToggleGroup";

describe("ToggleGroup", () => {
  const renderToggleGroup = (
    props: ToggleGroupProps = { type: "multiple" }
  ) => {
    return render(
      <ToggleGroup {...props}>
        <ToggleGroup.Item value="one" aria-label="Toggle one">
          One
        </ToggleGroup.Item>
        <ToggleGroup.Item value="two" aria-label="Toggle two">
          Two
        </ToggleGroup.Item>
        <ToggleGroup.Item value="three" aria-label="Toggle three">
          Three
        </ToggleGroup.Item>
      </ToggleGroup>
    );
  };

  it("renders a toggle group", () => {
    const { asFragment } = renderToggleGroup();
    const toggles = ["Toggle one", "Toggle two", "Toggle three"];

    toggles.forEach((toggle) =>
      expect(screen.getByRole("button", { name: toggle })).toHaveAttribute(
        "data-state",
        "off"
      )
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should toggle button state", async () => {
    const user = userEvent.setup();
    const { asFragment } = renderToggleGroup({
      type: "single",
      defaultValue: "two",
    });

    expect(screen.getByRole("radio", { name: "Toggle one" })).not.toBeChecked();
    expect(screen.getByRole("radio", { name: "Toggle two" })).toBeChecked();
    expect(
      screen.getByRole("radio", { name: "Toggle three" })
    ).not.toBeChecked();

    await user.click(screen.getByRole("radio", { name: "Toggle one" }));

    expect(screen.getByRole("radio", { name: "Toggle one" })).toBeChecked();
    expect(screen.getByRole("radio", { name: "Toggle two" })).not.toBeChecked();
    expect(
      screen.getByRole("radio", { name: "Toggle three" })
    ).not.toBeChecked();

    await user.click(screen.getByRole("radio", { name: "Toggle one" }));

    expect(screen.getByRole("radio", { name: "Toggle one" })).not.toBeChecked();
    expect(screen.getByRole("radio", { name: "Toggle two" })).not.toBeChecked();
    expect(
      screen.getByRole("radio", { name: "Toggle three" })
    ).not.toBeChecked();
    expect(asFragment()).toMatchSnapshot();
  });
});
