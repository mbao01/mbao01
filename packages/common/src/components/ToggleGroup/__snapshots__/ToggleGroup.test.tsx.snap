// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToggleGroup > renders a toggle group 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center gap-1"
    dir="ltr"
    role="group"
    style="outline: none;"
    tabindex="0"
  >
    <button
      aria-label="Toggle one"
      aria-pressed="false"
      class="btn disabled:text-opacity-50 min-h-fit h-10 rounded-md transition-colors disabled:pointer-events-none disabled:opacity-50 data-[state=off]:btn-ghost"
      data-radix-collection-item=""
      data-state="off"
      tabindex="-1"
      type="button"
    >
      One
    </button>
    <button
      aria-label="Toggle two"
      aria-pressed="false"
      class="btn disabled:text-opacity-50 min-h-fit h-10 rounded-md transition-colors disabled:pointer-events-none disabled:opacity-50 data-[state=off]:btn-ghost"
      data-radix-collection-item=""
      data-state="off"
      tabindex="-1"
      type="button"
    >
      Two
    </button>
    <button
      aria-label="Toggle three"
      aria-pressed="false"
      class="btn disabled:text-opacity-50 min-h-fit h-10 rounded-md transition-colors disabled:pointer-events-none disabled:opacity-50 data-[state=off]:btn-ghost"
      data-radix-collection-item=""
      data-state="off"
      tabindex="-1"
      type="button"
    >
      Three
    </button>
  </div>
</DocumentFragment>
`;

exports[`ToggleGroup > should toggle button state 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center gap-1"
    dir="ltr"
    role="group"
    style="outline: none;"
    tabindex="0"
  >
    <button
      aria-checked="false"
      aria-label="Toggle one"
      class="btn disabled:text-opacity-50 min-h-fit h-10 rounded-md transition-colors disabled:pointer-events-none disabled:opacity-50 data-[state=off]:btn-ghost"
      data-radix-collection-item=""
      data-state="off"
      role="radio"
      tabindex="0"
      type="button"
    >
      One
    </button>
    <button
      aria-checked="false"
      aria-label="Toggle two"
      class="btn disabled:text-opacity-50 min-h-fit h-10 rounded-md transition-colors disabled:pointer-events-none disabled:opacity-50 data-[state=off]:btn-ghost"
      data-radix-collection-item=""
      data-state="off"
      role="radio"
      tabindex="-1"
      type="button"
    >
      Two
    </button>
    <button
      aria-checked="false"
      aria-label="Toggle three"
      class="btn disabled:text-opacity-50 min-h-fit h-10 rounded-md transition-colors disabled:pointer-events-none disabled:opacity-50 data-[state=off]:btn-ghost"
      data-radix-collection-item=""
      data-state="off"
      role="radio"
      tabindex="-1"
      type="button"
    >
      Three
    </button>
  </div>
</DocumentFragment>
`;
